### PR TITLE
project is ready for code review , to portfolio project was added sea…

### PR DESCRIPTION
**Pull request by misterSokor**
…rch feature

### CODE REVIEW

File: src/main/java/com/store/litflix/repository/book/spec/TitleSpecificationProvider.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    @Override
+    public Specification<Book> getSpecification(String[] searchParameters) {
+        return new Specification<Book>() {
+            @Override
+            public Predicate toPredicate(Root<Book> root,
+                                         CriteriaQuery<?> query,
+                                         CriteriaBuilder criteriaBuilder) {
+                return root.get("title").in(Arrays.stream(searchParameters).toArray());
```
Comment (okuzan) [inline src/main/java/com/store/litflix/repository/book/spec/TitleSpecificationProvider.java]: extract repeated literal to constant

File: src/main/java/com/store/litflix/repository/book/BookSpecificationProviderManager.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+@RequiredArgsConstructor
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+    private final List<SpecificationProvider<Book>> bookSpecProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return bookSpecProviders.stream()
+                .filter(provider -> provider.getSearchKey().equals(key))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException(
```
Comment (okuzan) [inline src/main/java/com/store/litflix/repository/book/BookSpecificationProviderManager.java]: NoSuchElementException

File: src/main/java/com/store/litflix/repository/book/spec/AuthorSpecificationProvider.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    @Override
+    public Specification<Book> getSpecification(String[] searchParameters) {
+        return new Specification<Book>() {
+            @Override
+            public Predicate toPredicate(Root<Book> root,
+                                         CriteriaQuery<?> query,
+                                         CriteriaBuilder criteriaBuilder) {
+                return root.get("author").in(Arrays.stream(searchParameters).toArray());
```
Comment (okuzan) [inline src/main/java/com/store/litflix/repository/book/spec/AuthorSpecificationProvider.java]: same

**Review by okuzan — CHANGES_REQUESTED** [review]

**Comment by misterSokor** [comment]
add please to this commit message this change: [src/main/java/com/store/litflix/repository/book/BookSpecificationProviderManager.java](https://github.com/misterSokor/litflix-app/pull/1/files/b5c06cdca7296b28dd5528d56fbaabbd80056af6#diff-8ed5b47735956c616285b2b91ab00f28f461dcd7863a00bdcde4a13508a73252)
        return bookSpecProviders.stream()
                .filter(provider -> provider.getSearchKey().equals(key))
                .findFirst()
                .orElseThrow(() -> new RuntimeException(
@[okuzan](https://github.com/okuzan) okuzan [2 hours ago](https://github.com/misterSokor/litflix-app/pull/1#discussion_r1749876520)
NoSuchElementException

File: src/main/java/com/store/litflix/repository/book/BookSpecificationBuilder.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+@RequiredArgsConstructor
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    private final SpecificationProviderManager<Book> bookSpecificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParametersDto params) {
+        Specification<Book> spec = Specification.where(null);
+        if (params.author() != null && params.author().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager
+                    .getSpecificationProvider("author")
```
Comment (Elena-Bruyako) [inline src/main/java/com/store/litflix/repository/book/BookSpecificationBuilder.java]: "author" and "title" should be constants

File: src/main/java/com/store/litflix/repository/book/spec/AuthorSpecificationProvider.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Predicate;
+import jakarta.persistence.criteria.Root;
+import java.util.Arrays;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    private final String key = "author";
```
Comment (Elena-Bruyako) [inline src/main/java/com/store/litflix/repository/book/spec/AuthorSpecificationProvider.java]:
```suggestion
    private static final String AUTHOR_COLUMN_NAME = "author";
```

File: src/main/java/com/store/litflix/repository/book/spec/TitleSpecificationProvider.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Predicate;
+import jakarta.persistence.criteria.Root;
+import java.util.Arrays;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    private final String key = "title";
```
Comment (Elena-Bruyako) [inline src/main/java/com/store/litflix/repository/book/spec/TitleSpecificationProvider.java]: same

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**